Malmo/src/MissionSpec.cpp

```diff
@@ -18,656 +18,625 @@
+        drawing_decorator.add_child("DrawSphere", sphere);
     }
 
     void MissionSpec::drawLine(int x1, int y1, int z1, int x2, int y2, int z2, const std::string& blockType)
     {
-        DrawLine dl( blockType, x1, y1, z1, x2, y2, z2 );
-        DrawingDecorator dd;
-        dd.DrawObjectType().push_back(dl);
-        this->mission->ServerSection().ServerHandlers().DrawingDecorator().push_back( dd );
-        // TODO: improve the efficiency by adding to existing elements if present?
+        auto& drawing_decorator = getDrawingDecorator();
+        boost::property_tree::ptree sphere;
+        sphere.put("<xmlattr>.type", blockType);
+        sphere.put("<xmlattr>.x1", x1);
+        sphere.put("<xmlattr>.y1", y1);
+        sphere.put("<xmlattr>.z1", z1);
+        sphere.put("<xmlattr>.x2", x2);
+        sphere.put("<xmlattr>.y2", y2);
+        sphere.put("<xmlattr>.z2", z2);
+        drawing_decorator.add_child("DrawLine", sphere);
     }
     
     // ------------------ settings for the agents --------------------------------
 
     void MissionSpec::startAt(float x, float y, float z)
     {
-        this->mission->AgentSection().front().AgentStart().Placement() = PosAndDirection(x,y,z);
+        mission.put("Mission.AgentSection.AgentStart.Placement.<xmlattr>.x", x);
+        mission.put("Mission.AgentSection.AgentStart.Placement.<xmlattr>.y", y);
+        mission.put("Mission.AgentSection.AgentStart.Placement.<xmlattr>.z", z);
     }
 
     void MissionSpec::startAtWithPitchAndYaw(float x, float y, float z, float pitch, float yaw)
     {
-        PosAndDirection pos(x, y, z);
-        pos.pitch(pitch);
-        pos.yaw(yaw);
-        this->mission->AgentSection().front().AgentStart().Placement() = pos;
+        startAt(x, y, z);
+        mission.put("Mission.AgentSection.AgentStart.Placement.<xmlattr>.pitch", pitch);
+        mission.put("Mission.AgentSection.AgentStart.Placement.<xmlattr>.yaw", yaw);
     }
 
     void MissionSpec::endAt(float x, float y, float z, float tolerance)
     {
-        AgentHandlers::AgentQuitFromReachingPosition_optional& handler = this->mission->AgentSection().front().AgentHandlers().AgentQuitFromReachingPosition();
-        PointWithToleranceAndDescription p(x, y, z);
-        p.tolerance(tolerance);
-        if (!handler.present())
-        {
-            AgentQuitFromReachingPosition quitter;
-            quitter.Marker().push_back(p);
-            handler.set(quitter);
-        }
-        else
-        {
-            handler->Marker().push_back(p);
-        }
+        mission.put("Mission.AgentSection.AgentHandlers.AgentQuitFromReachingPosition.<xmlattr>.x", x);
+        mission.put("Mission.AgentSection.AgentHandlers.AgentQuitFromReachingPosition.<xmlattr>.y", y);
+        mission.put("Mission.AgentSection.AgentHandlers.AgentQuitFromReachingPosition.<xmlattr>.z", z);
+        mission.put("Mission.AgentSection.AgentHandlers.AgentQuitFromReachingPosition.<xmlattr>.tolerance", tolerance);
```

Comment:
Not sure this is right... AgentQuitFromReachingPosition contains a list of Markers, rather than a single x/y/z/tolerance.